diff --git a/node_modules/react-bulma-components/lib/components/utils.sass b/node_modules/react-bulma-components/lib/components/utils.sass
index b160d99..d246d52 100644
--- a/node_modules/react-bulma-components/lib/components/utils.sass
+++ b/node_modules/react-bulma-components/lib/components/utils.sass
@@ -1,6 +1,6 @@
 @charset "utf-8"
 
-@import "~_variables.sass"
+@import "./_variables.sass"
 @import "~bulma/sass/utilities/initial-variables.sass"
 @import "~bulma/sass/utilities/functions.sass"
 @import "~bulma/sass/utilities/derived-variables.sass"
